### PR TITLE
Fix the ability to shoot out of crates

### PR DESCRIPTION
**Pull request by Cojoke-dot**
## About the PR

bug
## Why / Balance

You could shoot out of crates while inside of them, making you unable to get hit by projectiles unless the crate was clicked on
## Technical details

adds a check to make sure that the shooter is not inside of a container whenever canceling collisions. For lasers, it checks to see if the last user is in a container in the same way as projectiles. (The bug got expanded halfway through making this)
## Media

https://github.com/space-wizards/space-station-14/assets/83733158/22170c70-5ce2-4d6f-ba6b-c559f8ee7dce

- [x] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

## Breaking changes

**Changelog**

:cl:
- fix: You can no longer shoot out of crates with guns

### CODE REVIEW

File: Content.Shared/Damage/Systems/RequireProjectileTargetSystem.cs [hunk 1]
```diff
@@ -26,7 +27,10 @@ private void PreventCollide(Entity<RequireProjectileTargetComponent> ent, ref Pr
         if (HasComp<ProjectileComponent>(other) &&
             CompOrNull<TargetedProjectileComponent>(other)?.Target != ent)
         {
-            args.Cancelled = true;
+            // Prevents shooting out of while inside of crates
+            var shooter = EnsureComp<ProjectileComponent>(other).Shooter;
+            if (CompOrNull<InsideEntityStorageComponent>(shooter)?.Storage != ent.Owner)
```
Comment (AJCM-git) [inline Content.Shared/Damage/Systems/RequireProjectileTargetSystem.cs:32]: use ContainerSystem.IsEntityOrParentInContainer

**Review by AJCM-git — CHANGES_REQUESTED** [review]

**Review by AJCM-git — CHANGES_REQUESTED** [review]
Please test things before requesting reviews

**Comment by AJCM-git** [comment]
> > Please test things before requesting reviews
> 
> Sorry about that
> 
> uh... here is a bug with how it works right now, I'm not sure how to fix it when using ContainerSystem.IsEntityOrParentInContainer. Its a minuscule bug that people are not going to see, but I will note its existence.
> 2024-06-13.21-59-12.mp4

You are right its minuscule because idk whats wrong in that video

**Comment by Cojoke-dot** [comment]
> You are right its minuscule because idk whats wrong in that video

The bullet hit the fax machine after I got in the crate (the second time) when I should have not hit. It just checks if the shooter is in a container, it does not differentiate the container and everything else.

**Review by AJCM-git — DISMISSED** [review]
you still have a pending review from sloth

**Review by metalgearsloth — CHANGES_REQUESTED** [review]
I haven't checked this out locally but I'm not really sure why this needs to also use RequireProjectileTarget changes to just add container checks?

**Comment by Cojoke-dot** [comment]
> I haven't checked this out locally but I'm not really sure why this needs to also use RequireProjectileTarget changes to just add container checks?

Is it not more efficient to never cancel it rather than the same checks for passover on the container too and cancel the PreventCollide? I may just not understand what your saying...

File: Content.Server/Weapons/Ranged/Systems/GunSystem.cs [hunk 2]
```diff
@@ -204,18 +206,20 @@ private void OnBallisticPrice(EntityUid uid, BallisticAmmoProviderComponent comp
...
-                            }
+                                    if (collide.HitEntity != gun.Target &&
+                                        CompOrNull<RequireProjectileTargetComponent>(collide.HitEntity)?.Active == true)
+                                    {
+                                        continue;
+                                    }
+
+                                    result = collide;
+                                    break;
+                                    }
```
Comment (metalgearsloth) [inline Content.Server/Weapons/Ranged/Systems/GunSystem.cs:222]: This indentation is cooked.

**Review by metalgearsloth — CHANGES_REQUESTED** [review]

**Pull request merged**